# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

In [2]:
#print current working directory
os.getcwd()

'C:\\Users\\tc18f\\Desktop\\springboard\\GuidedCapstone\\DataWrangling'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [3]:
df = pd.read_csv('step3_output.csv')
df.head()

Unnamed: 0                           Name       state  summit_elev  \
0           1            Eaglecrest Ski Area      Alaska         2600   
1           2               Hilltop Ski Area      Alaska         2090   
2           4            Sunrise Park Resort     Arizona        11100   
3           5  Yosemite Ski & Snowboard Area  California         7800   
4           8         Boreal Mountain Resort  California         7700   

   vertical_drop  trams  fastEight  fastSixes  fastQuads  quad  ...  \
0           1540      0        0.0          0          0     0  ...   
1            294      0        0.0          0          0     0  ...   
2           1800      0        0.0          0          1     2  ...   
3            600      0        0.0          0          0     0  ...   
4            500      0        0.0          0          1     1  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0              640.0       60.000000              45.0       44.0   
1               30.0       30.000000             150.0       36.0   
2              800.0       80.000000             115.0       49.0   
3               88.0      174.873239             110.0       84.0   
4              380.0      200.000000             150.0       54.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0            350.0          47.0      53.00000               90.0   
1             69.0          30.0      34.00000              152.0   
2            250.0          74.0      78.00000              104.0   
3            300.0          47.0      47.00000              107.0   
4            400.0          49.0      64.16681              150.0   

   NightSkiing_ac  clusters  
0             0.0         1  
1            30.0         1  
2            80.0         2  
3             0.0         0  
4           200.0         0  

[5 rows x 27 columns]

In [4]:
#getting rid off of 1st column
df = df.drop(columns='Unnamed: 0')
df.head()

Name       state  summit_elev  vertical_drop  \
0            Eaglecrest Ski Area      Alaska         2600           1540   
1               Hilltop Ski Area      Alaska         2090            294   
2            Sunrise Park Resort     Arizona        11100           1800   
3  Yosemite Ski & Snowboard Area  California         7800            600   
4         Boreal Mountain Resort  California         7700            500   

   trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0      0        0.0          0          0     0       0  ...   
1      0        0.0          0          0     0       1  ...   
2      0        0.0          0          1     2       3  ...   
3      0        0.0          0          0     0       1  ...   
4      0        0.0          0          1     1       3  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0              640.0       60.000000              45.0       44.0   
1               30.0       30.000000             150.0       36.0   
2              800.0       80.000000             115.0       49.0   
3               88.0      174.873239             110.0       84.0   
4              380.0      200.000000             150.0       54.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0            350.0          47.0      53.00000               90.0   
1             69.0          30.0      34.00000              152.0   
2            250.0          74.0      78.00000              104.0   
3            300.0          47.0      47.00000              107.0   
4            400.0          49.0      64.16681              150.0   

   NightSkiing_ac  clusters  
0             0.0         1  
1            30.0         1  
2            80.0         2  
3             0.0         0  
4           200.0         0  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [5]:
# select the state column
dfo = df[['state']]
df = pd.concat([df.drop(dfo, axis=1), pd.get_dummies(dfo)], axis=1)
df.head()

Name  summit_elev  vertical_drop  trams  \
0            Eaglecrest Ski Area         2600           1540      0   
1               Hilltop Ski Area         2090            294      0   
2            Sunrise Park Resort        11100           1800      0   
3  Yosemite Ski & Snowboard Area         7800            600      0   
4         Boreal Mountain Resort         7700            500      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  ...  \
0        0.0          0          0     0       0       4  ...   
1        0.0          0          0     0       1       0  ...   
2        0.0          0          1     2       3       1  ...   
3        0.0          0          0     0       1       3  ...   
4        0.0          0          1     1       3       1  ...   

   state_Rhode Island  state_South Dakota  state_Tennessee  state_Utah  \
0                   0                   0                0           0   
1                   0                   0                0           0   
2                   0                   0                0           0   
3                   0                   0                0           0   
4                   0                   0                0           0   

   state_Vermont  state_Virginia  state_Washington  state_West Virginia  \
0              0               0                 0                    0   
1              0               0                 0                    0   
2              0               0                 0                    0   
3              0               0                 0                    0   
4              0               0                 0                    0   

   state_Wisconsin  state_Wyoming  
0                0              0  
1                0              0  
2                0              0  
3                0              0  
4                0              0  

[5 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [6]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [7]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [8]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [9]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [10]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
from sklearn.metrics import explained_variance_score

# You can plug y_test and y_pred into the functions to evaluate the model
explained_variance_score(y_test, y_pred)

0.7246569140759163

In [11]:
#import mean_absolute_error
from sklearn.metrics import mean_absolute_error

# You can plug y_test and y_pred into the functions to evaluate the model
mean_absolute_error(y_test, y_pred)

6.3461828010822945

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [12]:
print(lm.intercept_)

59.288974235013725


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [13]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

Coefficient
summit_elev           1.309048e+00
vertical_drop         1.944617e+00
trams                 7.450933e+11
fastEight             1.250795e+13
fastSixes             8.808137e+13
fastQuads             1.227778e+12
quad                  1.277456e+12
triple                1.804066e+12
double                2.192899e+12
surface               1.956167e+12
total_chairs          4.136356e+12
Runs                  6.630859e-01
TerrainParks          1.097656e+00
LongestRun_mi         3.125000e-02
SkiableTerrain_ac     1.082031e+00
Snow Making_ac        1.367188e+00
daysOpenLastYear      1.038086e+00
yearsOpen             3.261719e-01
averageSnowfall       5.683594e-01
AdultWeekday          1.264062e+01
projectedDaysOpen     1.768555e+00
NightSkiing_ac        1.816406e-01
clusters              7.441406e-01
state_Alaska          2.222894e+12
state_Arizona         1.574797e+12
state_California      5.935868e+12
state_Colorado        4.894651e+12
state_Connecticut     3.494676e+12
state_Idaho           4.652534e+12
state_Illinois        3.131716e+12
state_Indiana         2.222894e+12
state_Iowa            2.717316e+12
state_Maine           4.394977e+12
state_Maryland        1.574797e+12
state_Massachusetts   4.894651e+12
state_Michigan        7.771101e+12
state_Minnesota       5.547964e+12
state_Missouri        2.222894e+12
state_Montana         5.123517e+12
state_Nevada          2.717316e+12
state_New Hampshire   5.547964e+12
state_New Jersey      1.574797e+12
state_New Mexico      4.119084e+12
state_New York        7.641771e+12
state_North Carolina  3.494676e+12
state_Ohio            3.494676e+12
state_Oregon          4.394977e+12
state_Pennsylvania    6.293987e+12
state_Rhode Island    1.574797e+12
state_South Dakota    2.222894e+12
state_Tennessee       1.574797e+12
state_Utah            4.394977e+12
state_Vermont         4.652534e+12
state_Virginia        2.717316e+12
state_Washington      4.394977e+12
state_West Virginia   3.131716e+12
state_Wisconsin       5.935868e+12
state_Wyoming         4.119084e+12

In [14]:
#sort the coefficients
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by=['Coefficient'], ascending=False)

Coefficient
fastSixes             8.808137e+13
fastEight             1.250795e+13
state_Michigan        7.771101e+12
state_New York        7.641771e+12
state_Pennsylvania    6.293987e+12
state_California      5.935868e+12
state_Wisconsin       5.935868e+12
state_Minnesota       5.547964e+12
state_New Hampshire   5.547964e+12
state_Montana         5.123517e+12
state_Colorado        4.894651e+12
state_Massachusetts   4.894651e+12
state_Idaho           4.652534e+12
state_Vermont         4.652534e+12
state_Oregon          4.394977e+12
state_Utah            4.394977e+12
state_Washington      4.394977e+12
state_Maine           4.394977e+12
total_chairs          4.136356e+12
state_New Mexico      4.119084e+12
state_Wyoming         4.119084e+12
state_Ohio            3.494676e+12
state_Connecticut     3.494676e+12
state_North Carolina  3.494676e+12
state_Illinois        3.131716e+12
state_West Virginia   3.131716e+12
state_Nevada          2.717316e+12
state_Iowa            2.717316e+12
state_Virginia        2.717316e+12
state_South Dakota    2.222894e+12
state_Indiana         2.222894e+12
state_Alaska          2.222894e+12
state_Missouri        2.222894e+12
double                2.192899e+12
surface               1.956167e+12
triple                1.804066e+12
state_Arizona         1.574797e+12
state_Rhode Island    1.574797e+12
state_New Jersey      1.574797e+12
state_Maryland        1.574797e+12
state_Tennessee       1.574797e+12
quad                  1.277456e+12
fastQuads             1.227778e+12
trams                 7.450933e+11
AdultWeekday          1.264062e+01
vertical_drop         1.944617e+00
projectedDaysOpen     1.768555e+00
Snow Making_ac        1.367188e+00
summit_elev           1.309048e+00
TerrainParks          1.097656e+00
SkiableTerrain_ac     1.082031e+00
daysOpenLastYear      1.038086e+00
clusters              7.441406e-01
Runs                  6.630859e-01
averageSnowfall       5.683594e-01
yearsOpen             3.261719e-01
NightSkiing_ac        1.816406e-01
LongestRun_mi         3.125000e-02

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [15]:
#changing the name to df1 and re-import and removing the columns
df1 = pd.read_csv('step3_output.csv')
df1 = df1.drop(columns=['Unnamed: 0', 'state'])
df1.head()

Name  summit_elev  vertical_drop  trams  \
0            Eaglecrest Ski Area         2600           1540      0   
1               Hilltop Ski Area         2090            294      0   
2            Sunrise Park Resort        11100           1800      0   
3  Yosemite Ski & Snowboard Area         7800            600      0   
4         Boreal Mountain Resort         7700            500      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  ...  \
0        0.0          0          0     0       0       4  ...   
1        0.0          0          0     0       1       0  ...   
2        0.0          0          1     2       3       1  ...   
3        0.0          0          0     0       1       3  ...   
4        0.0          0          1     1       3       1  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0              640.0       60.000000              45.0       44.0   
1               30.0       30.000000             150.0       36.0   
2              800.0       80.000000             115.0       49.0   
3               88.0      174.873239             110.0       84.0   
4              380.0      200.000000             150.0       54.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0            350.0          47.0      53.00000               90.0   
1             69.0          30.0      34.00000              152.0   
2            250.0          74.0      78.00000              104.0   
3            300.0          47.0      47.00000              107.0   
4            400.0          49.0      64.16681              150.0   

   NightSkiing_ac  clusters  
0             0.0         1  
1            30.0         1  
2            80.0         2  
3             0.0         0  
4           200.0         0  

[5 rows x 25 columns]

In [16]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X1 = df1.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y1 = df1['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler1 = preprocessing.StandardScaler().fit(X1)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled1=scaler1.transform(X1) 

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y1 = y1.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train1, X_test1, y_train1, y_test1 = train_test_split(X_scaled1, y1, test_size=0.25, random_state=1)

lm1 = linear_model.LinearRegression()
model1 = lm1.fit(X_train1,y_train1)

# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred1 = model1.predict(X_test1)

In [17]:
# You can plug y_test and y_pred into the functions to evaluate the model
print('explained variance score: ' + str(explained_variance_score(y_test1, y_pred1)))

# You can plug y_test and y_pred into the functions to evaluate the model
print('mean absolute error: ' + str(mean_absolute_error(y_test1, y_pred1)))

#The intercept is the mean AdultWeekend price for all the resorts given the other characteristics
print('intercept (mean AdultWeekend price): ' + str(lm1.intercept_))

explained variance score: 0.7082447599617352
mean absolute error: 6.631088537643005
intercept (mean AdultWeekend price): 59.36401134153611


In [18]:
#sorted coefficients for each variable
pd.DataFrame(abs(lm1.coef_), X1.columns, columns=['Coefficient']).sort_values(by=['Coefficient'], ascending=False)

Coefficient
AdultWeekday       1.303522e+01
vertical_drop      3.646044e+00
summit_elev        2.587142e+00
projectedDaysOpen  2.475929e+00
daysOpenLastYear   2.245973e+00
SkiableTerrain_ac  1.771206e+00
quad               1.349531e+00
Runs               1.026295e+00
Snow Making_ac     9.677264e-01
clusters           9.243694e-01
total_chairs       8.783973e-01
triple             8.521681e-01
TerrainParks       5.099411e-01
surface            4.981210e-01
double             4.489487e-01
fastQuads          2.947904e-01
LongestRun_mi      2.429989e-01
NightSkiing_ac     1.112098e-01
averageSnowfall    9.567883e-02
yearsOpen          7.693790e-02
fastSixes          6.661338e-16
fastEight          0.000000e+00
trams              0.000000e+00

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [19]:
#changing the name to df2 and re-import and removing the columns (base_elevs already dropped from previous step)
df2 = pd.read_csv('step3_output.csv')
df2 = df2.drop(columns=['Unnamed: 0', 'state', 'summit_elev'])
df2.head()

Name  vertical_drop  trams  fastEight  fastSixes  \
0            Eaglecrest Ski Area           1540      0        0.0          0   
1               Hilltop Ski Area            294      0        0.0          0   
2            Sunrise Park Resort           1800      0        0.0          0   
3  Yosemite Ski & Snowboard Area            600      0        0.0          0   
4         Boreal Mountain Resort            500      0        0.0          0   

   fastQuads  quad  triple  double  surface  ...  SkiableTerrain_ac  \
0          0     0       0       4        0  ...              640.0   
1          0     0       1       0        2  ...               30.0   
2          1     2       3       1        0  ...              800.0   
3          0     0       1       3        1  ...               88.0   
4          1     1       3       1        2  ...              380.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0       60.000000              45.0       44.0            350.0          47.0   
1       30.000000             150.0       36.0             69.0          30.0   
2       80.000000             115.0       49.0            250.0          74.0   
3      174.873239             110.0       84.0            300.0          47.0   
4      200.000000             150.0       54.0            400.0          49.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0      53.00000               90.0             0.0         1  
1      34.00000              152.0            30.0         1  
2      78.00000              104.0            80.0         2  
3      47.00000              107.0             0.0         0  
4      64.16681              150.0           200.0         0  

[5 rows x 24 columns]

In [20]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X2 = df2.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y2 = df2['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler2 = preprocessing.StandardScaler().fit(X2)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled2=scaler2.transform(X2) 

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y2 = y2.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train2, X_test2, y_train2, y_test2 = train_test_split(X_scaled2, y2, test_size=0.25, random_state=1)

lm2 = linear_model.LinearRegression()
model2 = lm2.fit(X_train2,y_train2)

# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred2 = model2.predict(X_test2)

In [21]:
# You can plug y_test and y_pred into the functions to evaluate the model
print('explained variance score: ' + str(explained_variance_score(y_test2, y_pred2)))

# You can plug y_test and y_pred into the functions to evaluate the model
print('mean absolute error: ' + str(mean_absolute_error(y_test2, y_pred2)))

#The intercept is the mean AdultWeekend price for all the resorts given the other characteristics
print('intercept (mean AdultWeekend price): ' + str(lm2.intercept_))

explained variance score: 0.7117761993497016
mean absolute error: 6.775107924056396
intercept (mean AdultWeekend price): 59.280020016248734


In [22]:
#sorted coefficients for each variable
pd.DataFrame(abs(lm2.coef_), X2.columns, columns=['Coefficient']).sort_values(by=['Coefficient'], ascending=False)

Coefficient
AdultWeekday       1.302285e+01
vertical_drop      2.588729e+00
projectedDaysOpen  2.448013e+00
SkiableTerrain_ac  2.372568e+00
daysOpenLastYear   2.306602e+00
Runs               1.454611e+00
quad               1.323059e+00
averageSnowfall    1.133890e+00
total_chairs       9.221221e-01
Snow Making_ac     8.777636e-01
fastQuads          7.130073e-01
triple             6.424451e-01
TerrainParks       6.067906e-01
clusters           5.642679e-01
LongestRun_mi      4.827329e-01
double             3.519952e-01
surface            3.487650e-01
NightSkiing_ac     2.507963e-01
yearsOpen          5.442500e-02
fastSixes          2.664535e-15
fastEight          2.664535e-15
trams              0.000000e+00

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.7247 | 6.3462 |-|
| Model 2. | 0.7082| 6.6311 |'state'|
| Model 3. | 0.7118 | 6.7751 |'state','summit_elev','base_elev'|

In [23]:
# You can plug y_test and y_pred into the functions to evaluate the model
print('explained variance score: ' + str(explained_variance_score(y_test, y_pred)))

# You can plug y_test and y_pred into the functions to evaluate the model
print('mean absolute error: ' + str(mean_absolute_error(y_test, y_pred)))

#The intercept is the mean AdultWeekend price for all the resorts given the other characteristics
print('intercept (mean AdultWeekend price): ' + str(lm.intercept_))
# You can plug y_test and y_pred into the functions to evaluate the model
print('explained variance score: ' + str(explained_variance_score(y_test1, y_pred1)))

# You can plug y_test and y_pred into the functions to evaluate the model
print('mean absolute error: ' + str(mean_absolute_error(y_test1, y_pred1)))

#The intercept is the mean AdultWeekend price for all the resorts given the other characteristics
print('intercept (mean AdultWeekend price): ' + str(lm1.intercept_))
# You can plug y_test and y_pred into the functions to evaluate the model
print('explained variance score: ' + str(explained_variance_score(y_test2, y_pred2)))

# You can plug y_test and y_pred into the functions to evaluate the model
print('mean absolute error: ' + str(mean_absolute_error(y_test2, y_pred2)))

#The intercept is the mean AdultWeekend price for all the resorts given the other characteristics
print('intercept (mean AdultWeekend price): ' + str(lm2.intercept_))

explained variance score: 0.7246569140759163
mean absolute error: 6.3461828010822945
intercept (mean AdultWeekend price): 59.288974235013725
explained variance score: 0.7082447599617352
mean absolute error: 6.631088537643005
intercept (mean AdultWeekend price): 59.36401134153611
explained variance score: 0.7117761993497016
mean absolute error: 6.775107924056396
intercept (mean AdultWeekend price): 59.280020016248734


Model Selection:

Different states will have different income level and thus states should matter, so I'll go with Model 1.